### PR TITLE
Stop the infinit loop for Resuable Blocks on the ID hook.

### LINKED ISSUE #869: Unique ID generation causing infinite loop when used inside reusable block
@Soare-Robert-Daniel If you add any block that uses the blockInit function inside a reusable block and you duplicate that reusable block on the same page, it's causing an infinite loop. It's more noticeable when you use Lottie block as it makes repeated HTTP requests.

### PR DESCRIPTION
**Pull request by Soare-Robert-Daniel**
fix #869 

@HardeepAsrani, my method for detecting reusable blocks is pretty brute. I did not find a more reliable way to check if the block from the page is a reusable block. So, for now, I am checking if the block ID is in one of the reusable blocks provided by 
```javascript
select( 'core/block-editor' )
			?.getSettings()
			?.__experimentalReusableBlocks.
```
----

#### Testing

Create a simple Lottie block, make it a reusable block, then insert the reusable block on the page a few times, and check if the animation is still showing. Also, the page must not freeze.

Also, try to add a new Otter block to it while is duplicated. For duplicated reusable blocks, the changes are reflected. If we add a new block inside it, change the colors then *publish* the page, the colors need to work in the frontend.

https://user-images.githubusercontent.com/17597852/162975660-085049aa-861c-4743-8851-6ee344a6ab12.mp4

### CODE REVIEW

File: src/blocks/helpers/block-utility.js [hunk 1]
```diff
@@ -147,7 +159,7 @@ export const addBlockId = ( args ) => {
 	 *
 	 * @type {Array.<string>}
 	 */
-	const blockIDs = window.themeisleGutenberg.blockIDs;
+	const blockIDs = window.themeisleGutenberg.blockIDs || [];
 
 	if ( attributes === undefined || setAttributes === undefined ) {
 		return ( savedId ) => {
```
Comment (Soare-Robert-Daniel) [inline src/blocks/helpers/block-utility.js:165]: @HardeepAsrani, for some unknown reasons, when I insert a Leaflet block, all the cleaning functions of the use effect Hook start looping for all the blocks in the reusable. If you insert `debugger;` here, then add a Leaflet block, the cycle will begin.

Comment (Soare-Robert-Daniel) [reply src/blocks/helpers/block-utility.js:165]: Or `console.count("clean");`

Comment (Soare-Robert-Daniel) [reply src/blocks/helpers/block-utility.js:165]: https://user-images.githubusercontent.com/17597852/163117640-8c814aa1-1761-4ac8-8e62-a69450fec337.mp4

**Comment by Soare-Robert-Daniel** [comment]
### Testing

#### Query

```javascript
new QueryQA().select('blocks').from("otter").where({ names: ['reusable-block-id-bug'] }).run()
```

Steps
1. Take the block from the query
2. Make it reusable
3. Insert the created reusable block, then duplicate it
4. Try to modify one of them
5. Check if the page does not freeze, and in fronted, everything is working.

⚠️ Some changes were made on the Query Engine. Some problems might appear until the CDN updated the source

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel Tested and found a single problem:

- [ ] When multiple Maps are added in the same post (being reusable blocks), only one of them will load in frontend and there's an error in console for the ones that don't load https://vertis.d.pr/mFl7lU

**Comment by Soare-Robert-Daniel** [comment]
@irinelenache, the problem should be solved now 🚀

**Comment by irinelenache** [comment]
@Soare-Robert-Daniel The issue is fixed now, thank you 🚀

**Pull request merged**